pyomo/contrib/pynumero/sparse/tests/test_block_matrix.py

```diff
@@ -142,39 +147,46 @@ def test_multiply(self):
         self.assertListEqual(res_dinopy.tolist(), res_scipy.tolist())
         self.assertListEqual(res_dinopy_flat.tolist(), res_scipy.tolist())
 
-        dense_mat = dinopy_mat.todense()
+        dense_mat = dinopy_mat.toarray()
         self.basic_m *= 5.0
-        self.assertTrue(np.allclose(dense_mat, self.basic_m.todense()))
+        self.assertTrue(np.allclose(dense_mat, self.basic_m.toarray()))
+
+    def test_mul_sparse_matrix(self):
+        m = self.basic_m
+
+        flat_prod = m.tocoo() * m.tocoo()
+        prod = m * m
 
-        flat_mat = self.basic_m.tocoo()
-        result = flat_mat * flat_mat
-        dense_result = result.toarray()
-        mat = self.basic_m * self.basic_m.tocoo()
-        dense_mat = mat.toarray()
-        self.assertTrue(np.allclose(dense_mat, dense_result))
+        self.assertIsInstance(prod, BlockMatrix)
+        self.assertTrue(np.allclose(flat_prod.toarray(), prod.toarray()))
 
-        # not supported block matrix times block matrix for now
-        #with self.assertRaises(Exception) as context:
-        #    mat = self.basic_m * self.basic_m.tocoo()
+        m2 = m.copy_structure()
+        ones = np.ones(m.shape)
+        m2.copyfrom(ones)
+        flat_prod = m.tocoo() * m2.tocoo()
+        prod = m * m2
+
+        self.assertIsInstance(prod, BlockMatrix)
+        self.assertTrue(np.allclose(flat_prod.toarray(), prod.toarray()))
```

Comment:
If we decide to support product with numpy array we should add a test here

Reply:
an easy way to support multiplication with 2d arrays would be to call the constructor of scipy.sparse on the array to get a "sparse copy" and use that sparse copy within our multiplication. This is important because some times scipy returns matrices of only one entry as 2d arrays.

Reply:
we should add a test anyway to make sure the correct error is thrown.